File: server/substrate/messages.py
from __future__ import annotations

from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, Field


# --- Client → Server ---


class ComputeRequest(BaseModel):
    type: Literal["compute_request"] = "compute_request"
    request_id: str
    node_id: str
    inputs: dict[str, Any] = Field(default_factory=dict)


class ConfigUpdateMsg(BaseModel):
    type: Literal["config_update"] = "config_update"
    node_id: str
    config: dict[str, Any]


class Resubscribe(BaseModel):
    type: Literal["resubscribe"] = "resubscribe"
    subscriptions: list[dict[str, str]]


ClientMessage = Annotated[
    Union[ComputeRequest, ConfigUpdateMsg, Resubscribe],
    Field(discriminator="type"),
]


# --- Server → Client ---


class GraphLoaded(BaseModel):
    type: Literal["graph_loaded"] = "graph_loaded"
    graph_id: str
    version: int
    nodes: list[dict[str, Any]]
    edges: list[dict[str, Any]]
    manifests: list[dict[str, Any]]


class NodeStateUpdated(BaseModel):
    type: Literal["node_state_updated"] = "node_state_updated"
    node_id: str
    data_patch: dict[str, Any]


class StreamEvent(BaseModel):
    type: Literal["stream_event"] = "stream_event"
    node_id: str
    stream: str
    cursor: str
    payload: dict[str, Any]
    ts: float


class ComputationResult(BaseModel):
    type: Literal["computation_result"] = "computation_result"
    request_id: str
    node_id: str
    ok: bool
    outputs: dict[str, Any] | None = None
    error: str | None = None


class ErrorMsg(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str
    node_id: str | None = None


class ReplayGap(BaseModel):
    type: Literal["replay_gap"] = "replay_gap"
    node_id: str
    stream: str
    requested_from: str
    earliest_available: str

File: frontend/src/types/messages.ts
export interface ComputeRequest {
  type: "compute_request";
  request_id: string;
  node_id: string;
  inputs: Record<string, unknown>;
}


export interface Resubscribe {
  type: "resubscribe";
  subscriptions: { node_id: string; stream: string }[];
}

export interface SubscribeWithResume {
  type: "subscribe_with_resume";
  subscriptions: { node_id: string; stream: string }[];
  last_ids: Record<string, string>;
}

export type ClientMessage = ComputeRequest | Resubscribe | SubscribeWithResume;

export interface GraphLoaded {
  type: "graph_loaded";
  graph_id: string;
  version: number;
  nodes: Record<string, unknown>[];
  edges: Record<string, unknown>[];
  manifests: Record<string, unknown>[];
}

export interface NodeStateUpdated {
  type: "node_state_updated";
  node_id: string;
  data_patch: Record<string, unknown>;
}

export interface StreamEvent {
  type: "stream_event";
  node_id: string;
  stream: string;
  cursor: string;
  payload: Record<string, unknown>;
  ts: number;
}

export interface ComputationResult {
  type: "computation_result";
  request_id: string;
  node_id: string;
  ok: boolean;
  outputs?: Record<string, unknown>;
  error?: string;
}

export interface ErrorMsg {
  type: "error";
  code: string;
  message: string;
  node_id?: string;
}

export interface ReplayGap {
  type: "replay_gap";
  stream: string;
  from_cursor: string;
  to_cursor: string;
  dropped: number;
}

export interface Resumed {
  type: "resumed";
  missed_count: number;
}

export type ServerMessage =
  | GraphLoaded
  | NodeStateUpdated
  | StreamEvent
  | ComputationResult
  | ErrorMsg
  | ReplayGap
  | Resumed;
